libp2p/stream_muxer/mplex/mplex_stream.py

```diff
@@ -90,63 +122,62 @@ def is_initiator(self) -> bool:
         but allows writing in the other direction.
-        :return: true if successful
         """
         # TODO error handling with timeout
-        # TODO understand better how mutexes are used from go repo
+
+        async with self.close_lock:
+            if self.event_local_closed.is_set():
+                return
+
         flag = (
             HeaderTags.CloseInitiator if self.is_initiator else HeaderTags.CloseReceiver
         )
+        # TODO: Raise when `mplex_conn.send_message` fails and `Mplex` isn't shutdown.
         await self.mplex_conn.send_message(flag, None, self.stream_id)
 
-        remote_lock = False
-        async with self.stream_lock:
-            if self.local_closed:
-                return True
-            self.local_closed = True
-            remote_lock = self.remote_closed
+        _is_remote_closed: bool
+        async with self.close_lock:
+            self.event_local_closed.set()
+            _is_remote_closed = self.event_remote_closed.is_set()
 
-        if remote_lock:
-            # FIXME: mplex_conn has no conn_lock!
-            async with self.mplex_conn.conn_lock:  # type: ignore
-                # FIXME: Don't access to buffers directly
-                self.mplex_conn.buffers.pop(self.stream_id)  # type: ignore
+        if _is_remote_closed:
+            # Both sides are closed, we can safely remove the buffer from the dict.
+            async with self.mplex_conn.streams_lock:
+                del self.mplex_conn.streams[self.stream_id]
 
-        return True
-
-    async def reset(self) -> bool:
+    async def reset(self) -> None:
         """
         closes both ends of the stream
         tells this remote side to hang up
-        :return: true if successful
         """
-        # TODO understand better how mutexes are used here
-        # TODO understand the difference between close and reset
-        async with self.stream_lock:
-            if self.remote_closed and self.local_closed:
-                return True
+        async with self.close_lock:
+            # Both sides have been closed. No need to event_reset.
+            if self.event_remote_closed.is_set() and self.event_local_closed.is_set():
+                return
+            if self.event_reset.is_set():
+                return
+            self.event_reset.set()
 
-            if not self.remote_closed:
+            if not self.event_remote_closed.is_set():
                 flag = (
                     HeaderTags.ResetInitiator
                     if self.is_initiator
                     else HeaderTags.ResetReceiver
                 )
-                await self.mplex_conn.send_message(flag, None, self.stream_id)
-
-            self.local_closed = True
-            self.remote_closed = True
+                asyncio.ensure_future(
+                    self.mplex_conn.send_message(flag, None, self.stream_id)
+                )
+                await asyncio.sleep(0)
```

Comment:
why do we need the `sleep` here?

Reply:
its not clear to me how it interacts w/ the `close_lock`

Reply:
yeah, i'd think the usual way to approach that would be to just `await self.mplex_conn.send_message` and then if we do need some type of synchronization we can add it, integrating the other parts of the machinery that need the `asyncio.Lock` or whatever mechanism makes sense